### PR TITLE
Fix incorrect CUDA `torch.nn.Embedding` result when max_norm is not None and indices are not sorted

### LINKED ISSUE #44792: Embedding max_norm wrong results on cuda when repeating indices are present
## 🐛 Bug

I need an reproducible output from model, but Embedding layer reproduces different results in some cases.

## To Reproduce

1. Set torch random seed
1. Use all of these options. Result will be reproducible if you miss or change any of them:
  * Use `cuda` device
  * Use `Embedding` layer with large `embedding_dim` and `max_norm` enabled
  * Get embeddings for large amount of repeated indexes.

Embeddings are different for different application runs.

```
import torch

torch.backends.cudnn.benchmark = False
torch.backends.cudnn.deterministic = True
torch.manual_seed(42)
torch.cuda.manual_seed_all(42)

device = torch.device('cuda:0')
model = torch.nn.Embedding(
            num_embeddings=2,
            embedding_dim=64,
            max_norm=1.0,
        ).to(device)
ix = torch.arange(2).long().to(device)
out = model(ix.repeat(2000))

for p in model.parameters():
    print((p ** 2).sum(dim=1, keepdim=True) ** 0.5)
print(out.sum())
```

## Expected behavior

I expect the same output for different application runs.

## Environment

```
Collecting environment information...
PyTorch version: 1.6.0
Is debug build: False
CUDA used to build PyTorch: 10.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.4 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.10.2

Python version: 3.7 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: Tesla P100-PCIE-16GB
GPU 1: Tesla P100-PCIE-16GB
GPU 2: Tesla P100-PCIE-16GB
GPU 3: Tesla P100-PCIE-16GB

Nvidia driver version: 435.21
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.1
[pip3] pytorch-ignite==0.4.0.post1
[pip3] torch==1.6.0
[pip3] torchvision==0.7.0
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               10.1.243             h6bb024c_0  
[conda] mkl                       2020.2                      256  
[conda] mkl-service               2.3.0            py37he904b0f_0  
[conda] mkl_fft                   1.1.0            py37h23d657b_0  
[conda] mkl_random                1.1.1            py37h0573a6f_0  
[conda] numpy                     1.19.1           py37hbc911f0_0  
[conda] numpy-base                1.19.1           py37hfa32c7d_0  
[conda] pytorch                   1.6.0           py3.7_cuda10.1.243_cudnn7.6.3_0    pytorch
[conda] pytorch-ignite            0.4.0.post1              pypi_0    pypi
[conda] torchvision               0.7.0                    pypi_0    pypi

```

## Additional context

cc @ezyang @gchanan @zou3519 @ngimel

### PR DESCRIPTION
**Pull request by kurtamohler**
Sorting indices before calling `thrust::unique` fixes the issue.
Fixes #44792

### CODE REVIEW

File: aten/src/ATen/native/cuda/Embedding.cu [hunk 1]
```diff
@@ -332,6 +332,8 @@ Tensor embedding_dense_backward_cuda(const Tensor & grad_, const Tensor & indice
 
 Tensor & embedding_renorm_cuda_(Tensor & self, const Tensor & indices,
                                 double max_norm, double norm_type) {
+  // See Note [Writing Nondeterministic Operations]
```
Comment (mruberry) [inline aten/src/ATen/native/cuda/Embedding.cu:335]: Also needs a comment explaining why the operation is nondeterministic

File: torch/testing/_internal/common_nn.py [hunk 2]
```diff
@@ -2364,6 +2364,17 @@ def fractional_max_pool3d_test(test_case):
         jacobian_input=False,
         check_gradgrad=False,
     ),
+    dict(
+        module_name='Embedding',
+        constructor_args=(4, 3, None, 1.0),
+        cpp_constructor_args='torch::nn::EmbeddingBagOptions(4, 3).max_norm(1.0)',
```
Comment (mruberry) [inline torch/testing/_internal/common_nn.py:2370]: Looks like this arg needs to be changed?

File: aten/src/ATen/native/cuda/Embedding.cu (hunk 1 above)
Comment (kurtamohler) [reply aten/src/ATen/native/cuda/Embedding.cu:335]: Right, I'm not sure why it's nondeterministic yet. I asked Natalia back in the issue if she knows.

**Comment by kurtamohler** [comment]
I think this is ready. The CI failures didn't seem to be my fault, but I'm rerunning them.

**Comment by mruberry** [comment]
So I'm curious about your thinking here, @kurtamohler,

- when duplicative entries are present on CUDA aren't the results just wrong? They may be nondeterministically wrong, but I thought the determinism flag was for algorithms which are correct-ish but whose values may vary (due to slight numerical instability, for example, or if there's a multiplicity of valid results).
- assuming this was a determinism issue for the moment, it's a determinism issue that only triggers when the input has a certain structure. Yet this PR suggests disallowing the behavior on all CUDA inputs, not just CUDA inputs with this property. If the determinism flag is set in cases like this should it do the extra work of reviewing the inputs' structure?

**Comment by ngimel** [comment]
@kurtamohler as the comment here says, thrust::unique does not work on unsorted unputs https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/Embedding.cu#L350-L356, so you should sort inputs before calling thrust on them. It may be extra work (if there are no repeating indices), but otherwise there's no way to guarantee that incorrect results won't be produced.
After the inputs are sorted, the behavior will no longer be nondeterministic, so it won't be necessary to set the flag.

**Comment by kurtamohler** [comment]
Thanks @ngimel! I've made that change and wrote a test based on the repro from the issue description.

File: test/test_nn.py [hunk 3]
```diff
@@ -2954,6 +2954,35 @@ def test_threshold_int(self):
...
+        # be able to get the same result by sorting the indices before
+        # giving them to Embedding. Then we can reorder the output
+        # to match the order of the unsorted indices
+        ix_sorted, ix_orig_indices = ix.sort()
+        ix_orig_indices = ix_orig_indices.tolist()
+        ix_reverse_sort_indices = [None] * len(ix_orig_indices)
+        for ind, orig_ind in enumerate(ix_orig_indices):
+            ix_reverse_sort_indices[orig_ind] = ind
+        out_presorted = create_embedding('cuda')(ix_sorted)[torch.tensor(ix_reverse_sort_indices, device='cuda')]
+        self.assertEqual(out, out_presorted)
```
Comment (kurtamohler) [inline test/test_nn.py:2984]: I'm not too sure if this test case is very important. I can remove it if not.

Comment (ngimel) [reply test/test_nn.py:2984]: Yeah, please remove it, just comparing results of inputs with repeating indices is enough.

Comment (kurtamohler) [reply test/test_nn.py]: Done

**Review by ngimel — APPROVED** [review]
